src/USER-MISC/angle_fourier_simple_approx.h

```diff
@@ -0,0 +1,157 @@
+/* -*- c++ -*- ----------------------------------------------------------
+   LAMMPS - Large-scale Atomic/Molecular Massively Parallel Simulator
+   https://www.lammps.org/, Sandia National Laboratories
+   Steve Plimpton, sjplimp@sandia.gov
+
+   Copyright (2003) Sandia Corporation.  Under the terms of Contract
+   DE-AC04-94AL85000 with Sandia Corporation, the U.S. Government retains
+   certain rights in this software.  This software is distributed under
+   the GNU General Public License.
+
+   See the README file in the top-level LAMMPS directory.
+------------------------------------------------------------------------- */
+
+/* ----------------------------------------------------------------------
+   Contributing author: Tim Bernhard (ETHZ)
+------------------------------------------------------------------------- */
+
+#ifdef ANGLE_CLASS
+// clang-format off
+AngleStyle(fourier/simple/approx,AngleFourierSimpleApprox);
+// clang-format on
+#else
+
+#ifndef LMP_ANGLE_FOURIER_SIMPLE_APPROX_H
+#define LMP_ANGLE_FOURIER_SIMPLE_APPROX_H
+
+#include "angle_fourier_simple.h"
+#include "math_const.h"
+#include <cassert>
+
+namespace LAMMPS_NS {
+
+/* ---------------------------------------------------------------------- */
+
+#define ONE_OVER_PI 0.318309886183790671538035357
+#define ONE_OVER_PI_SQUARE 0.1013211836423377714440499
+#define ONE_OVER_PI_CUBED 0.03225153443319948918450346
+#define TWO_PI_INVERSE 0.1591549430918953357690176
+
+/**
+ * Approximation of sine. |x*e(x)|≤2e-9
+ * @source M. Abramowitz and I. A. Stegun, Eds., Handbook of mathematical functions: with formulas, graphs, and mathematical tables, p. 43
+ */
+static float sinePolynomial(float x)
+{
+  const float x2 = x * x;
+
+  const float term1 = x2 * -0.0000000239 + 0.0000027526;
+  const float term2 = x2 * term1 - 0.0001984090;
+  const float term3 = x2 * term2 + 0.00833333315;
+  const float term4 = x2 * term3 - 0.16666666664;
+  return x * (1 + x2 * term4);
+}
+
+/**
+ * Approximation of sine
+ * @source M. Abramowitz and I. A. Stegun, Eds., Handbook of mathematical functions: with formulas, graphs, and mathematical tables, p. 43
+ */
+static float fastSin(float a)
+{
+  // wrap x within [0, TWO_PI)
+  // const float a = x * TWO_PI_INVERSE;
+  // x -= static_cast<int>(a) * MathConst::MY_2PI;
+  // if (x < 0.0f) x += MathConst::MY_2PI;
+  const float x = fmodf(MathConst::MY_2PI + fmodf(a, MathConst::MY_2PI), MathConst::MY_2PI);
+  // assert(x >= 0);
+  // assert(x < MathConst::MY_2PI);
+
+  // 4 pieces of hills: wrap x within [0, pi/2]
+  if (x < MathConst::MY_PI2)
+    return sinePolynomial(x);
+  else if (x < MathConst::MY_PI)
+    return sinePolynomial(MathConst::MY_PI - x);
+  else if (x < 3.0f * MathConst::MY_PI2)
+    return -sinePolynomial(x - MathConst::MY_PI);
+  else
+    return -sinePolynomial(MathConst::MY_2PI - x);
```

Comment:
The branching (and wrapping above) here is what makes this method slower than the Taylor series. Without it, it is faster (in a benchmark over 10'000'000 random, uniformly distributed angles).

Reply:
@GenieTim which compiler and which compiler flags do you use to compile LAMMPS?
Have you experimented with using '-O3 -ffast-math' (with GCC)?
or using the Intel compiler with '?-O3 -xHOST -ffast-math'?

I've experimented with your polynomial approximation for a quadrant and merged it with the other code I had found which does an approximation for `[-pi/2;pi/2]` and used some tricks to avoid if statements, yet made the code work on a wider range of arguments. I am attaching a small standalone test code that first fills arrays with the random inputs and results and later compares them so that the compiler cannot optimize the loops away. To get better statistics on timing, and avoid that memory bandwidth becomes a significant fact, the code allows to set the size of the array (neval) and then does repetitions (nrepeat) and you can set the interval for the arguments. The test code demonstrates that at full optimization single precision slows you down and gives reduced accuracy for larger arguments. But with full optimization, the Intel 2021.2 and GCC 11.1 compilers beat all approximation with their inline functions.
[fast-sin-tester.cpp.txt](https://github.com/lammps/lammps/files/6796078/fast-sin-tester.cpp.txt)

Reply:
For reference, I just uploaded the benchmark I was using to the following repository: https://github.com/GenieTim/cosine-benchmark
For compilation, I used the following: `clang++ main.cpp -std=c++20 -O3 -march=native -ffast-math -ftree-vectorize -Rpass-missed=loop-vectorize -Rpass-analysis=loop-vectorize`.

I highly appreciate your assistance, comments and the optimizations you found to wrap the ranges. Sorry that I cannot invest as much time as I would like to currently, I am confident though that my objective is going to be reached in not too long.

Reply:
Running your (@akohlmey) benchmark on Euler, compiled with `icpc -O3 -xHOST -ffast-math -std=c++11 -qopt-report=2 -qopt-report-phase=vec fast-sin-tester.cpp`, I get the following results when called with `./a.out 12 10000000 0 100` (though different parameters lead to comparable results): 
```
sin():  120000000 evaluations @ 2.16156e-05 us/call
sinf(): 120000000 evaluations @ 2.045e-08 us/call
poly_sinf(): 120000000 evaluations @ 0.0212622 us/call
poly_sin(): 120000000 evaluations @ 0.0049543 us/call
fast_sin(): 120000000 evaluations @ 0.00235397 us/all
```
Very notable is the fact that the c-standard `sin()` and `sinf()` seem to be faster than any approximations. I can replicate this actually with my cosine benchmark, but not with an equivalent acos benchmark. Looking at the assembly, icpc generates calls to `__svml_cos4_l9`, the Intel short vector math library. This is not the case with clang or gcc (naturally), where "native" cosine calls are not vectorized I believe. While these observations further support the use of  an approximation for acos, a cosine approximation would need to be used on non-intel compilers only. It is not clear to me yet why in LAMMPS itself my initial implementation was faster, but there are a multitude of explanations of course (daily load, statistical fluctuation, flags not applied/CMake misunderstood, Intel compiler did not do this optimization, etc.).
Question for the last possibility: while I know how to get my vectorization reports and assembly for one file only when I compile this one file only, how can you get those for one file/component only if you compile it into the whole (LAMMPS) project (is there some CMake magic I do not know of?)?

Reply:
@GenieTim Unlike the simple test code with simple and independent loops, most of the LAMMPS code will not easily vectorize, so it is rather meaningless to compare with benchmark numbers from compilations that have vectorization enabled. You have to look at the code in the INTEL package (and its rather convoluted code with preprocessor magic, vectorization directives and copying data to more vector friendly data structure) to see what kind of effort is required to get some meaningful vectorization out of the code.
GCC doesn't do quite as good a job as Intel in vectorization but can also dispatch to SVML.
But even without vectorization, the "native" GCC inline functions are about twice as fast as the best approximation (and with full accuracy) in my tests when using `-ffast-math`. That said, it may still be worth using the Intel compilers, since they also have their own math library, which is usually faster than with glibc offers. I've even has success with using the shared library version of it after the fact via LD_PRELOAD to speed up math heavy code (with lots of log() and exp() which are  more expensive than sin() or cos()) without recompiling.

I think the only additional thing to try out would be using the sincos() function since you need a sine and cosine of the same angle, IIRC. Toy codes have limited utility for modern compilers unless you explicitly turn off many optimizations and thus dumb them down enough so you can see numbers that are applicable to the real application at hand.

You can pass custom compiler flags to CMake for LAMMPS via `-DCMAKE_TUNE_FLAGS`

Reply:
FWIW, I rarely look at assembly code. Most of the time after the fact, e.g. via `perf report` after using `perf record` to get profiling data. Perf can "zoom in" and then uses objdump to disassemble the object code. When compiling with -g this is often useful enough for me to find out what I want to see (despite the debug info becoming rather unreliable with high optimization), since I mostly want to see whether serial or parallel SSE/AVX instructions are used and where the "hot spots" are.